### PR TITLE
Fix Issue 18219 - Private import inside struct leaks symbols when used as VarDeclaration types

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/dmd/dsymbol.d [hunk 1]
```diff
@@ -687,7 +687,7 @@ extern (C++) class Dsymbol : RootObject
      * Returns:
      *      symbol found, NULL if not
      */
-    final Dsymbol searchX(Loc loc, Scope* sc, RootObject id)
+    final Dsymbol searchX(Loc loc, Scope* sc, RootObject id, int flags)
```
Comment (JinShil) [inline src/dmd/dsymbol.d]: Please add `Params` DDoc comment.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (wilzbach) [reply src/dmd/dsymbol.d]: Normally DAutotest should check that and fail respectively (except that we broke it today - https://github.com/dlang/dmd/pull/7637)

**Comment by wilzbach** [comment]
Mind the CI failures:

```
dmd/parsetimevisitor.d(17): Error: no property 'Statement' for type 'ASTCodegen', did you mean 'dmd.statement.Statement'?
dmd/parsetimevisitor.d(19): Error: no property 'Expression' for type 'ASTCodegen', did you mean 'dmd.expression.Expression'?
dmd/parsetimevisitor.d(20): Error: no property 'TemplateParameter' for type 'ASTCodegen', did you mean 'dmd.dtemplate.TemplateParameter'?
dmd/parsetimevisitor.d(21): Error: no property 'Condition' for type 'ASTCodegen', did you mean 'dmd.cond.Condition'?
dmd/parsetimevisitor.d(22): Error: no property 'Initializer' for type 'ASTCodegen', did you mean 'dmd.init.Initializer'?
dmd/parsetimevisitor.d(26): Error: no property 'AliasThis' for type 'ASTCodegen', did you mean 'dmd.aliasthis.AliasThis'?
dmd/parsetimevisitor.d(27): Error: no property 'Declaration' for type 'ASTCodegen', did you mean 'dmd.declaration.Declaration'?
dmd/parsetimevisitor.d(28): Error: no property 'ScopeDsymbol' for type 'ASTCodegen', did you mean 'dmd.dsymbol.ScopeDsymbol'?
dmd/parsetimevisitor.d(29): Error: no property 'Import' for type 'ASTCodegen', did you mean 'dmd.dimport.Import'?
dmd/parsetimevisitor.d(30): Error: no property 'AttribDeclaration' for type 'ASTCodegen', did you mean 'dmd.attrib.AttribDeclaration'?
dmd/parsetimevisitor.d(31): Error: no property 'StaticAssert' for type 'ASTCodegen', did you mean 'dmd.staticassert.StaticAssert'?
dmd/parsetimevisitor.d(32): Error: no property 'DebugSymbol' for type 'ASTCodegen', did you mean 'dmd.dversion.DebugSymbol'?
dmd/parsetimevisitor.d(33): Error: no property 'VersionSymbol' for type 'ASTCodegen', did you mean 'dmd.dversion.VersionSymbol'?
dmd/parsetimevisitor.d(36): Error: no property 'Package' for type 'ASTCodegen', did you mean 'dmd.dmodule.Package'?
dmd/parsetimevisitor.d(37): Error: no property 'EnumDeclaration' for type 'ASTCodegen', did you mean 'dmd.denum.EnumDeclaration'?
dmd/parsetimevisitor.d(38): Error: no property 'AggregateDeclaration' for type 'ASTCodegen', did you mean 'dmd.aggregate.AggregateDeclaration'?
dmd/parsetimevisitor.d(39): Error: no property 'TemplateDeclaration' for type 'ASTCodegen', did you mean 'dmd.dtemplate.TemplateDeclaration'?
dmd/parsetimevisitor.d(40): Error: no property 'TemplateInstance' for type 'ASTCodegen', did you mean 'dmd.dtemplate.TemplateInstance'?
dmd/parsetimevisitor.d(41): Error: no property 'Nspace' for type 'ASTCodegen', did you mean 'dmd.nspace.Nspace'?
dmd/parsetimevisitor.d(45): Error: no property 'VarDeclaration' for type 'ASTCodegen', did you mean 'dmd.declaration.VarDeclaration'?
```

**Comment by RazvanN7** [comment]
@wilzbach Yeah, I need to talk with @andralex about that.

**Comment by wilzbach** [comment]
> Yeah, I need to talk with @andralex about that.

It looks similar to the bug found in https://github.com/dlang/phobos/pull/5584. I suspect this will need to trigger a deprecation for a while.

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/dsymbol.d]: Is this ok?

Comment (JinShil) [reply src/dmd/dsymbol.d]: > Is this ok?

IMO, yes.  Though I'm looking forward to replacing `flags` with a proper `enum`.  But that's another PR.

**Review by JinShil — DISMISSED** [review]

**Comment by RazvanN7** [comment]
After this is merged I sugest we rename `searchX` which does not say anything to `searchTypeMembers`
and move it to mtype.d since that is the only place where it is used. Later on, we should move all this functionality out of the AST nodes.

**Comment by RazvanN7** [comment]
Circle/ci and cybershadow fail because the version of the compiler used to compile the compiler is still under the effect of the lookup bug that this PR fixes

**Comment by wilzbach** [comment]
> still under the effect of the lookup bug that this PR fixes

Any way to work around it? You can bump CircleCi at `.circleci/run.sh`

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -2562,7 +2562,7 @@ extern (C++) abstract class Type : RootObject
             if (this != Type.terror)
             {
                 if (s)
-                    error(loc, "no property `%s` for type `%s`, did you mean `%s`?", ident.toChars(), toChars(), s.toChars());
+                    error(loc, "no property '%s' for type '%s', did you mean '%s'?", ident.toChars(), toChars(), s.toPrettyChars());
```
Comment (WalterBright) [inline src/dmd/mtype.d]: Please don't revert the back ticks.

**Comment by RazvanN7** [comment]
@andralex I transformed the error into a deprecation message.

**Review by andralex — APPROVED** [review]

**Comment by WalterBright** [comment]
@RazvanN7 this seems to have introduced a regression https://issues.dlang.org/show_bug.cgi?id=19103

**Pull request merged**